pkg/network/master/subnets.go

```diff
@@ -236,54 +233,59 @@ func (master *OsdnMaster) handleDeleteNode(obj interface{}) {
-					vnidInt, err := strconv.Atoi(vnid)
-					if err == nil && vnidInt >= 0 && uint32(vnidInt) <= network.MaxVNID {
-						hsAnnotations = make(map[string]string)
-						hsAnnotations[networkapi.FixedVNIDHostAnnotation] = strconv.Itoa(vnidInt)
-					} else {
-						log.Errorf("Vnid %s is an invalid value for annotation %s. Annotation will be ignored.", vnid, networkapi.FixedVNIDHostAnnotation)
-					}
-				}
-				_, err = master.addNode(name, hostIP, hsAnnotations, nil)
-				if err != nil {
-					log.Errorf("Error creating subnet for node %s, ip %s: %v", name, hostIP, err)
-					return nil
-				}
-			}
-		case cache.Deleted:
-			if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; !ok {
-				// release the subnet
-				_, ipnet, err := net.ParseCIDR(subnet)
-				if err != nil {
-					return fmt.Errorf("error parsing subnet %q for node %q for deletion: %v", subnet, name, err)
-				}
-				for _, possibleSubnetAllocator := range master.subnetAllocatorList {
-					possibleSubnetAllocator.ReleaseNetwork(ipnet)
-				}
-			}
+func (master *OsdnMaster) handleAddOrUpdateSubnet(obj, _ interface{}, eventType watch.EventType) {
+	hs := obj.(*networkapi.HostSubnet)
+	log.V(5).Infof("Watch %s event for HostSubnet %q", eventType, hs.Name)
+
+	if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; !ok {
+		return
+	}
+
+	// Delete the annotated hostsubnet and create a new one with an assigned subnet
+	// We do not update (instead of delete+create) because the watchSubnets on the nodes
+	// will skip the event if it finds that the hostsubnet has the same host
+	// And we cannot fix the watchSubnets code for node because it will break migration if
+	// nodes are upgraded after the master
+	err := master.networkClient.Network().HostSubnets().Delete(hs.Name, &metav1.DeleteOptions{})
+	if err != nil {
+		log.Errorf("Error in deleting annotated subnet from master, name: %s, ip %s: %v", hs.Name, hs.HostIP, err)
+		return
+	}
+	var hsAnnotations map[string]string
+	if vnid, ok := hs.Annotations[networkapi.FixedVNIDHostAnnotation]; ok {
+		vnidInt, err := strconv.Atoi(vnid)
+		if err == nil && vnidInt >= 0 && uint32(vnidInt) <= network.MaxVNID {
+			hsAnnotations = make(map[string]string)
+			hsAnnotations[networkapi.FixedVNIDHostAnnotation] = strconv.Itoa(vnidInt)
+		} else {
+			log.Errorf("VNID %s is an invalid value for annotation %s. Annotation will be ignored.", vnid, networkapi.FixedVNIDHostAnnotation)
 		}
-		return nil
-	})
+	}
+	_, err = master.addNode(hs.Name, hs.HostIP, hsAnnotations, nil)
+	if err != nil {
+		log.Errorf("Error creating subnet for node %s, ip %s: %v", hs.Name, hs.HostIP, err)
+		return
+	}
+}
+
+func (master *OsdnMaster) handleDeleteSubnet(obj interface{}) {
+	hs := obj.(*networkapi.HostSubnet)
+	log.V(5).Infof("Watch %s event for HostSubnet %q", watch.Deleted, hs.Name)
+
+	if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; !ok {
```

Comment:
Earlier we were releasing subnet when networkapi.AssignHostSubnetAnnotation is *not* present. (Refer: https://github.com/openshift/origin/blob/master/pkg/network/master/subnets.go#L276)
I'm guessing that's a existing bug in the code or my understanding could be wrong.
@rajatchopra can you confirm?

Reply:
Not a bug. We do not want to release if the annotation is present (because that indicates a manually created hostsubnet). Otherwise we do want to release it. Who will release it otherwise?

Reply:
Discussed with @rajatchopra on IRC,
`AssignHostSubnetAnnotation` is for F5 use case and there is no real node exists in the cluster.
HostSubnet is manually created with this special annotation to assign a subnet.
Since we don't get a node deletion event in this case, delete HostSubnet is not triggered. 
When user manually deletes this created HostSubnet and if the special annotation is present, then we need to release the subnet (existing bug).
I will add some comments to make it clear.